core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -694,7 +694,7 @@ private[storage] class PartiallyUnrolledIterator[T](
   }
 
   override def next(): T = {
-    if (unrolled == null) {
+    if (unrolled == null || !unrolled.hasNext) {
```

Comment:
Yeah that's a fair point because it should be fair to put the burden on the caller to check `hasNext` before calling `next` and now `TorrentBroadcast` does that. However, are there other call sites that need that type of fix too? if all callers are well behaved then I agree we could revert the added `hasNext` call in `next`.